include/NAS2D/MathUtils.h

```diff
@@ -29,5 +30,26 @@ namespace NAS2D
 
 	int divideUp(int to_divide, int divisor);
 
+	
+
+/**
+ * \fn T convertDomainToRange(T a, T b)
+ *
+ * Applies a linear conversion to map an input domain to an output range for built-in types.
+ *
+ * \param	value: The value to be converted from the input domain to the output range.
+ * \param	minDomain: Smallest value of the domain.
+ * \param	maxDomain: Largest value of the domain.
+ * \param	minRange: Smallest value of the range.
+ * \param	maxRange: Largest value of the range.
+ *
+ * \return	Returns the mapped value in the closed range [minOutputRange,maxOutputRange]
+ */
+template<typename T>
+T convertDomainToRange(const T& value, const T& minDomain, const T& maxDomain, const T& minRange, const T& maxRange)
```

Comment:
I would recommend two template parameters here, one for `InputType` and one for `OutputType`. The first three parameters would be `InputType`, while the last two would be `OutputType`. That way you could convert a scalar value to say a 2D vector.

----

Input and Output types are largely going to be simple primitive types, or simple value-type structs. We should consider passing by value rather than `const &`. For those who are curious what impact that has on the code, I highly recommend checking https://godbolt.org/ to see what assembly code is generated by a few mainstream compilers.

Reply:
I would disagree on not using `const&`. See: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#fcall-parameter-passing

Specifically the center of the table (emphasis mine):

>
>Don't know (e.g., unfamiliar type, **template**) input parameters: `f(const&)`

![image](https://user-images.githubusercontent.com/9336881/73549784-e6c02280-4408-11ea-923b-31cfe4816176.png)